flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/reuse/SinkReuser.java

```diff
@@ -0,0 +1,213 @@
+    }
+
+    private Set<Sink> reuseSinkAndAddUnion(List<ReusableSinkGroup> reusableSinkGroups) {
+        final Set<Sink> reusedSinkNodes = Collections.newSetFromMap(new IdentityHashMap<>());
+        reusableSinkGroups.forEach(
+                group -> {
+                    List<Sink> originalSinks = group.originalSinks;
+                    if (originalSinks.size() <= 1) {
+                        Preconditions.checkState(originalSinks.size() == 1);
+                        reusedSinkNodes.add(originalSinks.get(0));
+                        return;
+                    }
+                    List<RelNode> allSinkInputs = new ArrayList<>();
+                    for (Sink sinkNode : originalSinks) {
+                        allSinkInputs.add(sinkNode.getInput());
+                    }
+                    // Use the first sink node as the final reused sink node
+                    Sink reusedSink = originalSinks.get(0);
+
+                    Union unionForReusedSinks;
+
+                    unionForReusedSinks =
+                            new BatchPhysicalUnion(
+                                    reusedSink.getCluster(),
+                                    group.inputTraitSet,
+                                    allSinkInputs,
+                                    true,
+                                    // use sink input row type
+                                    reusedSink.getRowType());
+
+                    reusedSink.replaceInput(0, unionForReusedSinks);
+                    reusedSinkNodes.add(reusedSink);
+                });
+        return reusedSinkNodes;
+    }
+
+    /**
+     * Grouping sinks that can be reused with each other.
+     *
+     * @param allSinkNodes in the plan.
+     * @return a list contains all grouped sink.
+     */
+    private List<ReusableSinkGroup> groupReusableSink(List<Sink> allSinkNodes) {
+        List<ReusableSinkGroup> reusableSinkGroups = new ArrayList<>();
+
+        for (Sink currentSinkNode : allSinkNodes) {
+            boolean canBeReused = false;
+            String currentSinkDigest = getDigest(currentSinkNode);
+            SinkAbilitySpec[] currentSinkSpecs;
+
+            currentSinkSpecs = ((BatchPhysicalSink) currentSinkNode).abilitySpecs();
+            RelTraitSet currentInputTraitSet = currentSinkNode.getInput().getTraitSet();
+            for (ReusableSinkGroup group : reusableSinkGroups) {
+                // Only table sink with the same digest, specs and input trait set can be reused
+                if (!(group.digest.equals(currentSinkDigest)
+                        && Arrays.equals(group.sinkAbilitySpecs, currentSinkSpecs)
+                        && group.inputTraitSet.equals(currentInputTraitSet))) {
+                    continue;
+                }
+
+                canBeReused = true;
+                group.originalSinks.add(currentSinkNode);
+                break;
+            }
+
+            if (!canBeReused) {
```

Comment:
What does this if mean? It is saying that if we cannot Reuse then add to a resuableSinkGroup. This does not look right. I was expecting `if (canBeReused) {`

Reply:
`canBeReused == false` means we need to create a new ReusableSinkGroup for this sink node.